backup/moodle2/restore_applaunch_stepslib.php

```diff
@@ -0,0 +1,160 @@
+        }
+
+        // Return the paths wrapped into standard activity structure.
+        return $this->prepare_activity_structure($paths);
+    }
+
+    /**
+     * Process the main activity instance.
+     *
+     * @param $data
+     * @throws coding_exception
+     */
+    protected function process_applaunch($data) {
+        $data = (object) $data;
+        $data->apptypeid = 0; // Set the app type id to empty. It will be updated once the apptype is processed.
+        $data = $this->filter_default_persistent_fields($data);
+        $newinstance = new \mod_applaunch\applaunch(0, $data);
+        $newinstance->save();
+        $this->apply_activity_instance($newinstance->get('id'));
+    }
+
+    /**
+     * Process the app type.
+     *
+     * @param $data
+     * @throws coding_exception
+     */
+    protected function process_apptype($data) {
+        $data = (object) $data;
+        $apptype = $this->get_existing_app_type($data);
+        if (empty($apptype)) {
+            // No existing type found, so create a new one.
+            $data = $this->filter_default_persistent_fields($data);
+            $apptype = new app_type(0, $data);
+            $apptype->save();
+        }
+        // Update applaunch instance.
+        $applaunch = new \mod_applaunch\applaunch($this->get_new_parentid('applaunch'));
+        $applaunch->set('apptypeid', $apptype->get('id'));
+        $applaunch->save();
+    }
+
+    /**
+     * Process the user completion.
+     *
+     * @param $data
+     */
+    protected function process_completion($data) {
+        $data = (object) $data;
+
+        // Get the new user id and cmid.
+        $data->userid = $this->get_mappingid('user', $data->userid);
+        $data->cmid = $this->task->get_moduleid();
+
+        // Create new completion record.
+        $data = $this->filter_default_persistent_fields($data);
+        $newcompletion = new \mod_applaunch\completion(0, $data);
+        $newcompletion->save();
+    }
+
+    /**
+     * Get the existing app type if it exists.
+     *
+     * @param $data
+     * @return persistent|null
+     * @throws restore_step_exception
+     */
+    protected function get_existing_app_type($data): ?persistent {
+        // Check if the app type is already mapped.
+        if ($apptypeid = $this->get_mappingid('apptype', $data->id)) {
+            return new app_type($apptypeid);
+        }
+
+        // If same site, try and find app type by id.
+        $apptype = app_type::get_record(['id' => $data->id]);
+        if ($this->task->is_samesite() && $apptype !== false) {
+            // If found, map it.
+            $this->set_mapping('apptype', $data->id, $data->id);
+            return $apptype;
+        }
+
+        // Try and find an existing app type matching name, description and url. If found, we assume it's the same app type.
+        $apptype = app_type::get_record([
```

Comment:
This seems pretty brittle, could we perhaps just match on URL only? that is the actual part that needs to be 1:1, and would allow for different/customised wording between multiple same client systems for example

Reply:
In my head, I was trying to take into account more features being added to the app_type but even if that was the case, the backup/restore should be updated at that point. 

Also added mapping the found apptype id that I missed.